src/nanothread.cpp

```diff
@@ -63,6 +65,65 @@ struct Worker {
 
 static Pool *pool_default_inst = nullptr;
 static std::mutex pool_default_lock;
+static uint32_t cached_core_count = 0;
+
+uint32_t core_count() {
+    // assumes atomic word size memory access
+    if (cached_core_count)
+        return cached_core_count;
+
+#if defined(__linux__)
+    /* Determine the number of present cores */
+    uint32_t ncores = sysconf(_SC_NPROCESSORS_CONF);
+
+    /* Don't try to query CPU affinity if running inside Valgrind */
+    if (getenv("VALGRIND_OPTS") == NULL) {
+        /* Some of the cores may not be available to the user
+           (e.g. on certain cluster nodes) -- determine the number
+           of actual available cores here. */
+        uint32_t ncores_logical = ncores;
+        size_t size = 0;
+        cpu_set_t *cpuset = NULL;
+        int retval = 0;
+
+        /* The kernel may expect a larger cpu_set_t than would
+           be warranted by the physical core count. Keep querying
+           with increasingly larger buffers if the
+           pthread_getaffinity_np operation fails */
+        for (uint32_t i = 0; i < 10; ++i) {
+            size = CPU_ALLOC_SIZE(ncores_logical);
+            cpuset = CPU_ALLOC(ncores_logical);
+            if (!cpuset) {
```

Comment:
This function is exposed through a C-level API, and the default assumption there for some platforms like on Windows is that functions cannot throw an exception (see the `NANOTHREAD_THROW` macro that is used to annotate a few). My suggestion would be that this function is implemented to _never_ throw, and that it falls back to `std::thread::hardware_concurrency` in case of a failure.

Remove unnecessary braces, return `0` to signal a failure.

Reply:
Yes makes sense, I wasn't convinced by the exceptions either and just using the hardware_concurrency as a default makes sense.